File: 0054-spiral-matrix/0054-spiral-matrix.js
/**
 * @param {number[][]} matrix
 * @return {number[]}
 */
var spiralOrder = function(matrix) {
    let res = [];
    let left = 0;
    let top = 0;
    let right = matrix[0].length; //number of columns
    let bottom = matrix.length; //number of rows

    while(left < right && top < bottom){
        
        //top row
        for(let i = left; i < right; i++){
            res.push(matrix[top][i]);
        }
        top = top + 1;
        
        //rightmost column
        for(let i = top; i < bottom; i++){
            res.push(matrix[i][right - 1]);
        }
        right = right - 1;
        
        if(!(left < right && top < bottom)){
            break;
        }
        
        //bottom row
        for(let i = right - 1; i >= left; i--){
            res.push(matrix[bottom - 1][i]);
        }
        bottom = bottom - 1;
        
        //first colums
        for(let i = bottom - 1; i >= top; i--){
            res.push(matrix[i][left]);
        }
        left = left + 1;
    }

    return res;
};

File: 0054-spiral-matrix/0054-spiral-matrix.py
class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        left, top = 0, 0
        right = len(matrix[0])
        bottom = len(matrix)
        result = []
        
        while left < right and top < bottom:
            #traverse the topmost row
            for i in range(left, right):
                result.append(matrix[top][i])
            
            top+=1
            
            #traverse last column
            for i in range(top, bottom):
                result.append(matrix[i][right - 1])
            
            right-=1
            
            #if matrix is a single row
            if not (left < right and top < bottom):
                break
            
            for i in range(right - 1, left - 1, -1):
                result.append(matrix[bottom - 1][i])
                
            bottom-=1
            
            for i in range(bottom - 1, top - 1, -1):
                result.append(matrix[i][left])
            
            left+=1
        
        return result